### PR TITLE
Fix overflow for counter metric

### PR DESCRIPTION
**Pull request by cijothomas**
Overflow handling for Counter fixed to:
1. Allow 2000 unique entries, *excluding* the overflow and 0 attributes case. i.e we will store 2002 entries, specially treating the 0 attribute and overflow to optimize performance (later)
2. Fix the error handling to be triggered once per collection interval at max.
3. Avoid unnecessary cloning of overflow atttribute.
4. Added tests and perf tests to have targeted testing for overflow scenario.

Note: Did not touch Histogram,Gauge, as there are no tests today for them, and their overflow logic looks incorrectly implemented. Will comeback to them after Counter is fully resolved.

### CODE REVIEW

**Review by lalitb — APPROVED** [review]
LGTM. Thanks.

File: opentelemetry-sdk/src/metrics/internal/sum.rs [hunk 1]
```diff
@@ -55,12 +55,12 @@ impl<T: Number<T>> ValueMap<T> {
...
+                        *val += measurement;
+                        return;
                     } else {
-                        values
-                            .entry(STREAM_OVERFLOW_ATTRIBUTE_SET.clone())
-                            .and_modify(|val| *val += measurement)
-                            .or_insert(measurement);
-                        global::handle_error(MetricsError::Other("Warning: Maximum data points for metric stream exceeded. Entry added to overflow.".into()));
+                        values.insert(STREAM_OVERFLOW_ATTRIBUTE_SET.clone(), measurement);
+                        global::handle_error(MetricsError::Other("Warning: Maximum data points for metric stream exceeded. Entry added to overflow. Further overflows to same metric until next collect will not be logged.".into()));
```
Comment (utpilla) [inline opentelemetry-sdk/src/metrics/internal/sum.rs:63]: > Further overflows to same metric until next collect will not be logged."

Is it because we clear the `HashMap` during every collect?

Comment (utpilla) [inline opentelemetry-sdk/src/metrics/internal/sum.rs:63]:
```suggestion
                        global::handle_error(MetricsError::Other("Warning: Maximum data points for metric stream exceeded. Entry added to overflow. Subsequent overflows to same metric until next collect will not be logged.".into()));
```

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 2]
```diff
@@ -161,6 +161,35 @@ mod tests {
...
 
+    #[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+    async fn counter_overflow_delta() {
+        // Arrange
+        let mut test_context = TestContext::new(Temporality::Delta);
+        let counter = test_context.u64_counter("test", "my_counter", None);
+
+        // Act
+        // Record measurements with A:0, A:1,.......A:1999, which just fits in the 2000 limit
+        for v in 0..2000 {
```
Comment (utpilla) [inline opentelemetry-sdk/src/metrics/mod.rs:172]: It would be good to also test the overflow scenario when a measurement with zero tags has also been recorded.

File: stress/src/metrics_overflow.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+fn main() {
+    for v in 0..2001 {
+        COUNTER.add(100, &[KeyValue::new("A", v.to_string())]);
+    }
+    throughput::test_throughput(test_counter);
+}
+
+fn test_counter() {
+    COUNTER.add(1, &[KeyValue::new("A", "2001")]);
```
Comment (utpilla) [inline stress/src/metrics_overflow.rs:42]: I think we should use random `ATTRIBUTE_VALUES` here to have a fair comparison with the regular stress test for metrics. Emitting the same dimension here would always give better results than this: https://github.com/open-telemetry/opentelemetry-rust/blob/aca4b52e20ee6728c91eb74744adfffa68e061d6/stress/src/metrics.rs#L2-L6

File: opentelemetry-sdk/src/metrics/internal/sum.rs (hunk 1 above)
Comment (cijothomas) [reply opentelemetry-sdk/src/metrics/internal/sum.rs:63]: yes.

File: opentelemetry-sdk/src/metrics/mod.rs (hunk 2 above)
Comment (cijothomas) [reply opentelemetry-sdk/src/metrics/mod.rs:172]: yes, good point. (we'd be expecting 2000+1+1 = 2002 points). Will add next.

File: stress/src/metrics_overflow.rs (hunk 3 above)
Comment (cijothomas) [reply stress/src/metrics_overflow.rs:42]: The goal for this test was to make sure that in the event of some rogue input, the process won't crash due to metrics i.e are we keeping our memory under cap? are we doing excessive logging (like we were doing prior to this PR) ? etc.

I think this test can do the below to simulate that?
`COUNTER.add(1, &[KeyValue::new("A", i++)]);`

Will keep iterating on this.

Comment (utpilla) [reply stress/src/metrics_overflow.rs:42]: We could remove `ATTRIBUTE_VALUES` if we are not using them.
